Refactors HTTPRoute Filters

danehans opened the pull request:
Previously, filtering was performed after a `match` rule and before an `action` rule. This restricts the ability to perform filter-like modifications to a request on a target-by-target basis. Additionally, the meaning of `filter` means to "remove" something and the name should be changed to properly reflect the ability to remove or add something to the matched request. This PR a) renames `filter` to `modify` and b) pluralizes `action` and c) moves the `modify` (previously `filter`) struct from `HTTPRouteRule` to `HTTPRequestAction`.

Fixes https://github.com/kubernetes-sigs/service-apis/issues/55

/assign @bowei 
/cc @robscott @hbagdi

danehans commented:
Partially fixes #58.

hbagdi commented on api/v1alpha1/httproute_types.go:
```diff
@@ -264,6 +264,14 @@ type ForwardToTarget struct {
+	// Support: Core
+	//
+	// +optional
+	Filter *HTTPRouteFilter `json:"filter" protobuf:"bytes,3,opt,name=filter"`
```
We should consider renaming `HTTPRouteFilter` to `HTTPRequestFilter` or make it as simple as `HTTPFilter`, which can be used as an extension point to add in routing functionality like caching, authentication, complex transformations, etc.

yiyangy replied:
There can also be filter to add/remove headers from the response prior to sending the response to the client. Shall that be extension?

hbagdi replied:
@yiyangy Great timing. I was actually thinking about this right now and was considering opening an issue to discuss exactly this. Keeping it in extension seems reasonable to me.

danehans replied:
@yiyangy thanks for the review and for bringing response filtering to my attention. I would like to separate request and response handling, so please create an issue and link to this PR.

danehans replied:
@hbagdi commit `53eaea7` replaces `Route` with `Request` in `HTTPRoute` match, filter and action types. I decided on `Request` based on possibly supporting `Response` types in the future.

hbagdi commented:
Other than the minor comment, this looks good.
/lgtm

hbagdi commented on api/v1alpha1/httproute_types.go:
```diff
@@ -72,15 +72,25 @@ type HTTPRouteHost struct {
-	Match *HTTPRouteMatch `json:"match" protobuf:"bytes,1,opt,name=match"`
-	// Filter defines what filters are applied to the request.
+	Match *HTTPRequestMatch `json:"match" protobuf:"bytes,1,opt,name=match"`
+	// Filter defines a schema for applying filters to a matched HTTP request.
```
The doc comment for filter in the target is more specific and helpful than this one:

> // Filter defines filters to apply to the request before forwarding it to
	// a targetRef.


Can we also add the semantic meaning of the filter in this godoc?
Essentially, saying that the filters MUST be executed when this route is selected for a request.

hbagdi commented on api/v1alpha1/httproute_types.go:
```diff
@@ -264,6 +276,14 @@ type ForwardToTarget struct {
 	TargetPort *TargetPort `json:"targetPort" protobuf:"bytes,2,opt,name=targetPort"`
+
+	// Filter defines filters to apply to the request before forwarding it to
+	// a targetRef.
```
The intention here is that these filters are executed in addition to the filters defined at the Route-level. Can we include that in this description to avoid confusion down the road?

danehans commented:
@hbagdi commit `414cf7b` addresses your feedback. I have changed the scope of the PR, so please reread https://github.com/kubernetes-sigs/service-apis/pull/179#issue-414386223 and let me know your thoughts.

hbagdi commented:
I preferred the older `Filter` over the new `Modifier` name. The reason being that this narrows down the scope of things that can be included in to only be able to perform request manipulation. 
Filter was broad enough that proxies could use it as an extension point to implement custom functionality for cross-cutting concerns around logging, tracing, authentication, authroziation, etc as well.

danehans commented:
> The reason being that this narrows down the scope of things that can be included in to only be able to perform request manipulation.

@hbagdi thanks for the feedback. I struggle using `Filter` to mean modify or manipulate something because filter, by definition, means to remove. @yiyangy previously addressed the need to perform response manipulation. If a `Filter` is specific to manipulating a request, does it not apply to response filtering? Aside from the renaming, what are your thoughts on the PR moving `Filter/Modify` to `HTTPRequestAction` and pluralizing `Action` so that filtering/modifying is performed on a target-by-target basis?

hbagdi commented:
This was briefly explained in office hours on May 27th and here is the summary:

The reason I preferred `Filter` above `Modifier` was to use that as an extension point to implement a variety of functionality that can work during the request life-cycle as well as response life-cycle.
As an example, Kong can run a variety of [plugins](https://docs.konghq.com/hub/) before forwarding the request to the upstream destination and after the response is received from the upstream. These are usually used implement authentication, logging, metrics, and other custom policies like rate-limiting, transformations, CORS handling etc.

While my example is specific to Kong, most proxies follow a general pattern:
- accept a request
- match it to a specific route/location
- run custom functionality
- forward the request upstream
- same during the response life-cycle

hbagdi commented:
@danehans  Do we need anything from this PR that is not in #249?

hbagdi commented:
ping @danehans

robscott commented:
As discussed in office hours, closing for now.

/close
